src/rust/engine/src/downloads.rs

```diff
@@ -26,30 +28,40 @@ struct NetDownload {
 
 impl NetDownload {
   async fn start(core: &Arc<Core>, url: Url, file_name: String) -> Result<NetDownload, String> {
-    // TODO: Retry failures
-    let response = core
+    let try_download = || async {
+      core
       .http_client
       .get(url.clone())
       .send()
       .await
-      .map_err(|err| format!("Error downloading file: {}", err))?;
+      .map_err(|err| (format!("Error downloading file: {}", err), true))
+      .and_then(|res|
+        // Handle common HTTP errors.
+        if res.status().is_server_error() {
+          Err((format!(
+            "Server error ({}) downloading file {} from {}",
+            res.status().as_str(),
+            file_name,
+            url,
+          ), true))
+        } else if res.status().is_client_error() {
+          Err((format!(
+            "Client error ({}) downloading file {} from {}",
+            res.status().as_str(),
+            file_name,
+            url,
+          ), false))
+        } else {
+          Ok(res)
+        })
+    };
+
+    // TODO: Allow the retry strategy to be configurable?
+    let retry_strategy = ExponentialBackoff::from_millis(10).map(jitter).take(3);
+    let response = RetryIf::spawn(retry_strategy, try_download, |err: &(String, bool)| err.1)
+      .await
+      .map_err(|(err, _)| err)?;
```

Comment:
As mentioned: I think it's totally fine to TODO making it configurable, but the error message should likely be enriched to give some information about the retries: i.e.:
```suggestion
      .map_err(|(err, _)| format!("After {num_attempts} attempts: {err}"))?;
```
...so that the next person who comes along has a clear hint of where to go looking if they want to add configurability.

Reply:
I agree this would be nice but I'm not sure we have access to the true `num_attempts` 🤔 if we hit a 4xx error we won't retry at all.

Instead of modifying this error, what do you think of adding logging to the `try_download` branches? So we can log something like "hit 5xx error, retrying" or "hit 4xx error, not retrying"

Reply:
Looks like `RetryIf::spawn` only cares that the 1st arg implements `IntoIterator` with `Duration` items - I could write a thin wrapper around `ExponentialBackoff` that counts/logs the retry attempts as part of `next()`

Reply:
> I agree this would be nice but I'm not sure we have access to the true `num_attempts` (I think it's hidden away inside the `RetryIf::spawn` implementation) 🤔 if we hit a 4xx error we won't retry at all.

I just meant extracting the constant "`3`" (or `4`) from the code above, and then using it in two places.

Reply:
Ooooh. I see what you meant here. Was missing the `RetryIf` aspect of this. Hm, it's a bit awkward to lose the conditional retry for client errors... between the two, not reporting the number of retries would be preferable? Sorry for the pivot.